command.go

```diff
@@ -1066,37 +1066,39 @@ func (cmd *XMessageSliceCmd) readReply(rd *proto.Reader) error {
 	return nil
 }
 
+// xMessageParser implements proto.MultiBulkParse.
+func xMessageParser(rd *proto.Reader, n int64) (interface{}, error) {
+	id, err := rd.ReadString()
+	if err != nil {
+		return nil, err
+	}
+
+	var values map[string]interface{}
+
+	v, err := rd.ReadArrayReply(stringInterfaceMapParser)
+	if err != nil {
+		if err != proto.Nil {
+			return nil, err
+		}
+	} else {
+		values = v.(map[string]interface{})
+	}
+
+	return XMessage{
+		ID:     id,
+		Values: values,
+	}, nil
+}
+
 // xMessageSliceParser implements proto.MultiBulkParse.
 func xMessageSliceParser(rd *proto.Reader, n int64) (interface{}, error) {
 	msgs := make([]XMessage, n)
 	for i := 0; i < len(msgs); i++ {
-		i := i
-		_, err := rd.ReadArrayReply(func(rd *proto.Reader, n int64) (interface{}, error) {
-			id, err := rd.ReadString()
-			if err != nil {
-				return nil, err
-			}
-
-			var values map[string]interface{}
-
-			v, err := rd.ReadArrayReply(stringInterfaceMapParser)
-			if err != nil {
-				if err != proto.Nil {
-					return nil, err
-				}
-			} else {
-				values = v.(map[string]interface{})
-			}
-
-			msgs[i] = XMessage{
-				ID:     id,
-				Values: values,
-			}
-			return nil, nil
-		})
+		msg, err := rd.ReadArrayReply(xMessageParser)
 		if err != nil {
 			return nil, err
 		}
+		msgs[i] = msg.(XMessage)
```

Comment:
It was written this way to avoid interface{} allocation and `XMessage` copying. It is important in this loop since we could read hundreds of messages in one request.

I suggest to refactor this code as:

```go
_, err := rd.ReadArrayReply(func(rd *proto.Reader, n int64) (interface{}, error) {
    var err error
    msgs[i], err = readMessage(rd)
    if err != nil {
        return nil, err
    }
    return nil, nil
})
```